### ISSUE TITLE
Ruby 2.2 warnings

### ISSUE THREAD

**Issue by mhenrixon** [rspec/rspec-mocks#933]
I get some warnings running rspec 3.2.1 on ruby 2.2.2 not sure if this is being handled already but these are the warnings

> rspec-mocks-3.2.1/lib/rspec/mocks/argument_list_matcher.rb:84: warning: Comparable#== will no more rescue exceptions of #<=> in the next release.
> rspec-mocks-3.2.1/lib/rspec/mocks/argument_list_matcher.rb:84: warning: Return nil in #<=> if the comparison is inappropriate or avoid such comparison

Lines: 67, 84 & 87

Also get a warning on fuzzy matcher with the same information:
> rspec-support-3.2.2/lib/rspec/support/fuzzy_matcher.rb:15 

This seems to be the solution: https://github.com/rdoc/rdoc/commit/23c276de 
More info here: https://github.com/rdoc/rdoc/issues/331

**Comment by myronmarston** [comment]
Thanks for reporting this!  I've not seen those warnings before.  Can you provide some example code that reproduces these warnings?

Actually, looking at those lines of code you referenced, they are lines where we call `==` on user objects, so I believe the warning is likely triggered by an object you've defined, not anything internal to RSpec.  We made changes to avoid calling `==` on user objects in some situations (since it may be defined in such a way that it raises an error) in #924 and rspec/rspec-support#193.  I believe that if you update to rspec HEAD you may see the warnings go away (although if the warnings are due to your objects you should probably consider fixing the definition of `<=>` to satisfy the warnings).

**Comment by myronmarston** [comment]
Closing for lack of a response (and because I believe we've probably already solved this as mentioned above).  Will gladly reopen if @mhenrixon provides an example that demonstrates it's still an issue.

**Comment by mhenrixon** [comment]
Sorry @myronmarston I've just completely missed the previous commend you made. I'll do some tracking this weekend but something tells me this was already fixed for me somehow.

**Comment by mhenrixon** [comment]
@myronmarston do you have any gist of what is needed in terms of dependencies to use head for all of rspec? Was going to try some stuff out but keep running into broken dependencies in the current project.

**Comment by mhenrixon** [comment]
No more warnings after running all the tests with HEAD so this one is totally ok to close! :)

**Comment by myronmarston** [comment]
Glad you figured that out.  For future reference, we've put instructions for this in our READMEs:

https://github.com/rspec/rspec-rails#installation

**Comment by mhenrixon** [comment]
> https://github.com/rspec/rspec-rails#installation

Perfect! Thanks a bunch

**Comment by shepmaster** [comment]
I continue to see these warnings with Ruby 2.2.3 and RSpec Expectations 3.2.1. Do we believe that the fixes mentioned above have been released?

```
rspec-expectations-3.2.1/lib/rspec/matchers/built_in/include.rb:95: warning: Return nil in #<=> if the comparison is inappropriate or avoid such comparison.
rspec-expectations-3.2.1/lib/rspec/matchers/built_in/include.rb:95: warning: Comparable#== will no more rescue exceptions of #<=> in the next release.
```

**Comment by myronmarston** [comment]
The fixes are in RSpec 3.3 so please consider upgrading.

**Comment by shepmaster** [comment]
> RSpec 3.3 so please consider upgrading.

Ugh, of course. I did actually *try* upgrading, but I had other gems preventing upgrading to 3.3. Thanks for being patient with user error!

**Comment by shepmaster** [comment]
Hmm. Still there. I guess I need to do more diving to get to the bottom of it:

```
rspec-expectations-3.3.1/lib/rspec/matchers/built_in/include.rb:111: warning: Comparable#== will no more rescue exceptions of #<=> in the next release.
rspec-expectations-3.3.1/lib/rspec/matchers/built_in/include.rb:111: warning: Return nil in #<=> if the comparison is inappropriate or avoid such comparison
```

**Comment by shepmaster** [comment]
Ok, so the problem is due to something similar fixed in #924. I had code that boiled down to:

```ruby
require 'rspec'

class Foo
  include Comparable

  def initialize(a_unique_name)
    @a_unique_name = a_unique_name
  end

  def <=>(other)
    # Note implicit assumption that `other` has the `a_unique_name` method
    @a_unique_name == other.a_unique_name
    # But `AnyArgMatcher` doesn't have that method!
  end
end

RSpec.describe Foo do
  it do
    items = [Foo.new('hi'), Foo.new('bye')]
    expect(items).to include(anything, anything, anything)
  end
end
```

In my case, the fix is to be more defensive in the `<=>` method.

**Comment by JonRowe** [comment]
Hmm so the warning was generated by your code but appeared to be coming from ours? I'm not sure what we can do to help with that
